homeassistant/components/sum/sensor.py

```diff
@@ -0,0 +1,214 @@
+    result = []
+    for _, sensor_value in sensor_values:
+        if sensor_value in [STATE_UNKNOWN, STATE_UNAVAILABLE]:
+            continue
+        result.append(sensor_value)
+
+    value: float = round(sum(result), round_digits)
+    return value
+
+
+class SumSensor(SensorEntity):
+    """Representation of a sum sensor."""
+
+    _attr_icon = ICON
+    _attr_should_poll = False
+    _attr_state_class = SensorStateClass.MEASUREMENT
+
+    def __init__(
+        self,
+        entity_ids: list[str],
+        name: str,
+        round_digits: float,
+        unique_id: str | None,
+    ) -> None:
+        """Initialize the sum sensor."""
+        self._attr_unique_id = unique_id
+        self._entity_ids = entity_ids
+        self._round_digits = int(round_digits)
+
+        self._attr_name = name
+
+        self.sum: float | None = None
+        self.valid_entities: list[str] | None = None
+        self._unit_of_measurement = None
+        self.states: dict[str, Any] = {}
+
+    async def async_added_to_hass(self) -> None:
+        """Handle added to Hass."""
+        self.async_on_remove(
+            async_track_state_change_event(
+                self.hass, self._entity_ids, self._async_sum_sensor_state_listener
+            )
+        )
+
+        # Replay current state of source entities
+        for entity_id in self._entity_ids:
+            state = self.hass.states.get(entity_id)
+            state_event = Event("", {"entity_id": entity_id, "new_state": state})
+            self._async_sum_sensor_state_listener(state_event, update_state=False)
+
+        self._calc_values()
+
+    @property
+    def native_value(self) -> float | None:
+        """Return the state of the sensor."""
+        return self.sum
+
+    @property
+    def extra_state_attributes(self) -> Mapping[str, Any] | None:
+        """Return extra state attributes."""
+        return {
+            ATTR_ENTITIES: self._entity_ids,
+            ATTR_VALID_ENTITIES: self.valid_entities,
+            ATTR_COUNT_VALID: len(self.valid_entities) if self.valid_entities else 0,
+        }
+
+    @callback
+    def _async_sum_sensor_state_listener(
+        self, event: EventType, update_state: bool = True
+    ) -> None:
+        """Handle the sensor state changes."""
+        new_state: State | None = event.data.get("new_state")
+        entity: str = event.data["entity_id"]
+
+        if (
+            new_state is None
+            or new_state.state is None
+            or new_state.state
+            in [
+                STATE_UNKNOWN,
+                STATE_UNAVAILABLE,
+            ]
+        ):
+            self.states[entity] = STATE_UNKNOWN
+            if not update_state:
+                return
+
+            self._calc_values()
+            self.async_write_ha_state()
+            return
+
+        if self._unit_of_measurement is None:
+            self._unit_of_measurement = new_state.attributes.get(
+                ATTR_UNIT_OF_MEASUREMENT
+            )
```

Comment:
This is odd thing to do. If I sum two sensors with different units, the unit will jump between those depending on which one last updated.